Flow: add loki.source.file component (#2546)
grafana/agent: docs/sources/flow/reference/components/loki.source.file.md

---
aliases:
- /docs/agent/latest/flow/reference/components/loki.source.file
title: loki.source.file
---

# loki.source.file

`loki.source.file` reads log entries from files and forwards them to other
`loki.*` components.

Multiple `loki.source.file` components can be specified by giving them
different labels.

## Usage

```river
loki.source.file "LABEL" {
  targets    = TARGET_LIST
  forward_to = RECEIVER_LIST
}
```

## Arguments
The component starts a new reader for each of the given `targets` and fans out
log entries to the list of receivers passed in `forward_to`.

> The list of arguments that can be used to configure the block is presented
> below.

Reviewer: ```suggestion

`loki.source.file` supports the following arguments:
```

Reviewer: Suggestion for simplifying and making this consistent with other component Arguments sections.

* `loki_source_file_read_lines_total` (counter): Number of lines read.
* `loki_source_file_encoding_failures_total` (counter): Number of encoding failures.
* `loki_source_file_files_active_total` (gauge): Number of active files.

## Component behavior
Each element in the list of `targets` as a set of key-value pairs called
_labels_.
The set of targets can either be _static_, or dynamically provided periodically
by a service discovery component. The special label `__path__` _must always_ be
present and must point to the absolute path of the file to read from.
<!-- TODO(@tpaschalis) refer to discovery.fileglob -->

The `__path__` value is  available as the `filename` label to each log entry
the component reads. All other labels starting with a double underscore are
considered _internal_ and are removed from the log entries before they're
passed to other `loki.*` components.

The component uses its data path (a directory named after the domain's
fully qualified name) to store its _positions file_. The positions file is used
to store read offsets, so that in case of a component or Agent restart,
`loki.source.file` can pick up tailing from the same spot. 

> In case a file is removed from the `targets` list, its positions file entry
> is also removed; that means that when it's added back on, `loki.source.file`
> starts reading it from the beginning.

Reviewer: ```suggestion
If a file is removed from the `targets` list, its positions file entry
is also removed. When it's added back on, `loki.source.file`
starts reading it from the beginning.
```

---
aliases:
- /docs/agent/latest/flow/reference/components/loki.source.file
title: loki.source.file
---

# loki.source.file

`loki.source.file` reads log entries from files and forwards them to other
`loki.*` components.

Multiple `loki.source.file` components can be specified by giving them
different labels.

## Usage

```river
loki.source.file "LABEL" {
  targets    = TARGET_LIST
  forward_to = RECEIVER_LIST
}
```

## Arguments
The component starts a new reader for each of the given `targets` and fans out
log entries to the list of receivers passed in `forward_to`.
The list of arguments that can be used to configure the block is presented
below.

Name         | Type                   | Description          | Default | Required
------------ | ---------------------- | -------------------- | ------- | --------
`targets`    | `list(map(string))`    | List of files to read from. | | yes

> `forward_to` | `list(chan loki.Entry)` | List of receivers to send log entries to. | | yes

Author: Should we alias this to a more generic type name like `LogsReceiver`? Or leave it as-is for now?

Reviewer 2: I'd say use something generic to reduce the chances that someone asks what something like `chan loki.Entry` means.

## Blocks

The `loki.source.file` component doesn't support any inner blocks and is
configured fully through arguments.

## Exported fields

`loki.source.file` does not export any fields.

## Component health

`loki.source.file` is only reported as unhealthy if given an invalid
configuration.

## Debug information

`loki.source.file` does not expose any component-specific debug information. ????

## Component behavior
Each element in the list of `targets` as a set of key-value pairs called
_labels_.
The set of targets can either be _static_, or dynamically provided periodically

> by a service discovery component such as `discovery.fileglob`. The special

Author: The fileglob component does not exist yet (but will exist by the next release). Should I remove the example and add it back when it's ready?

Reviewer 2: IMO yes, I'd prefer avoiding referencing unimplemented components in our documentation even if we'll add them before the release (some people may still be reading the `next` version documentation)

of the file to read from.

The `__path__` value will be available as the `filename` label to each log
entry the component will read. All other labels starting with a double
underscore are considered _internal_ and will be removed from the log entries
after they've been read.

The component will use its data path (a directory named after the domain's
fully qualified name) to store its _positions file_. The positions file is used
to store read offsets, so that in case of a component or Agent restart,
`loki.source.file` can pick up tailing from the same spot. 

In case a file is removed from the `targets` list, its positions file entry
is also removed; that means that when it's added back on, `loki.source.file`
will start reading from the beginning.

## Example

This example collects log entries from the files specified in the targets
argument and forwards them to a `loki.write` component so they are can be 
written to Loki.

```river
loki.source.file "tmpfiles" {
  targets    = [
    {__path__ = "/tmp/foo.txt", "color" = "pink"},
    {__path__ = "/tmp/bar.txt", "color" = "blue"},
    {__path__ = "/tmp/baz.txt", "color" = "grey"},
  ] 
  forward_to = [loki.write.local.receiver]
}

> loki.write "local" {

Author: The component does not exist yet (but will exist by the next release). Should I remove the example and add it back when it's ready?

Author: I've removed it and added a TODO in an HTML comment to add it back once it's merged.
